sns_ik_lib/src/sns_acc_ik_base.cpp

```diff
@@ -216,6 +216,108 @@ SnsIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen::V
+    ROS_ERROR("Primary goal did not find a solution! Terminating..");
+    return exitCode;
+  }
+
+  //--- find the solution for the secondary goal
+
+  *taskScaleCS = 1.0;  // task scale (assume feasible solution until proven otherwise)
+  Eigen::MatrixXd I = Eigen::MatrixXd::Identity(getNrOfJoints(), getNrOfJoints());
+
+  // Keep track of which joints are saturated:
+  std::vector<bool> jointIsFree(getNrOfJoints(), true);
+
+  /*
+   * W is a diagonal selection matrix which indicates free joints.
+   * The entry of 1 indicates the corresponding joint is free for the task,
+   * and 0 indicates the corresponding joint is saturated.
+   */
+  Eigen::MatrixXd W = I;  // null-space selection matrix
+  for (size_t jntIdx = 0; jntIdx < getNrOfJoints(); jntIdx++) {
+    if (ddq1(jntIdx) > (getUpperBounds())(jntIdx) + BOUND_TOLERANCE ||
+        ddq1(jntIdx) < (getLowerBounds())(jntIdx) - BOUND_TOLERANCE) {
+          W(jntIdx, jntIdx) = 0.0;
+          jointIsFree[jntIdx] = false;
+    }
+  }
+
+  // Compute nullspace projection matrix
+  Eigen::MatrixXd Jinv, Pinv;
+  if(!pinv(J, &Jinv, LIN_SOLVE_RESIDUAL_TOL)) {
+    ROS_ERROR("Pseudo-inverse of J cannot be computed!");
+    return ExitCode::InternalError;
+  }
+
+  Eigen::MatrixXd P1 = I - Jinv*J; // for primary task
+  if(!pinv((I - W)*P1, &Pinv, LIN_SOLVE_RESIDUAL_TOL)){
+    // if (I-W) is a zero matrix, inverse is the same
+    Pinv = (I - W)*P1;
+  }
+  Eigen::MatrixXd Pcs = (I - Pinv)*P1; // for both primary and joint saturation tasks
+
+  // Compute "a" and "b" from the paper.   (J*W*a = dx)
+  Eigen::VectorXd a = Pcs * ddqCS;
+  Eigen::ArrayXd b = ddq1.array();
+
+  // Compute the task scale associated with each joint
+  Eigen::ArrayXd jntScaleFactorArr(getNrOfJoints());
+  Eigen::ArrayXd lowMargin = (getLowerBounds() - b);
+  Eigen::ArrayXd uppMargin = (getUpperBounds() - b);
+  for (size_t i = 0; i < getNrOfJoints(); i++) {
+    if (jointIsFree[i]) {
+      jntScaleFactorArr(i) = findScaleFactor(lowMargin(i), uppMargin(i), a(i));
+    } else {  // joint is constrained
+      jntScaleFactorArr(i) = POS_INF;
+    }
+  }
+
+  // Compute the most critical scale factor and corresponding joint index
+  *taskScaleCS = jntScaleFactorArr(0);  // minimum value of jntScaleFactorArr()
+  for (size_t i = 1; i < getNrOfJoints(); i++) {
+    if (jntScaleFactorArr(i) < *taskScaleCS) {
+      *taskScaleCS = jntScaleFactorArr(i);
+    }
+  }
+
+  if (*taskScaleCS == POS_INF) {
+    // if all joints are saturated, secondary goal becomes infeasible!
+    *taskScaleCS = 0;
+    ROS_WARN("All joints are saturated! Secondary goal is infeasible!");
+  }
+  else if (*taskScaleCS > 1.0) {
+    ROS_ERROR("Task scale is %f, which is more than 1.0", *taskScaleCS);
+    return ExitCode::InternalError;
+  }
+  else if (*taskScaleCS < MINIMUM_FINITE_SCALE_FACTOR) {
+    ROS_WARN("Secondary goal is infeasible! scaling --> zero");
+  }
+
+  // compute the additional joint velocity due to the secondary goal
```

Comment:
is the additional joint acceleration for SnsAccIkBase ?

Reply:
Yes, good catch, I will fix the comment.